db/pull.go

```diff
@@ -0,0 +1,226 @@
+	}
+	newHead := makeSnapshot(db.noms, genesis.Ref(), pullResp.StateID, db.noms.WriteValue(patchedMap.NomsMap()), patchedMap.NomsChecksum(), pullResp.LastMutationID)
+	db.noms.SetHead(db.noms.GetDataset(LOCAL_DATASET), db.noms.WriteValue(marshal.MustMarshal(db.noms, newHead)))
+
+	return pullResp.ClientViewInfo, db.init()
+}
+
+type puller interface {
+	Pull(noms types.ValueReadWriter, baseState Commit, url string, clientViewAuth string, clientID string) (Commit, servetypes.ClientViewInfo, error)
+}
+
+type defaultPuller struct{}
+
+func (defaultPuller) Pull(noms types.ValueReadWriter, baseState Commit, url string, clientViewAuth string, clientID string) (Commit, servetypes.ClientViewInfo, error) {
+	return pull(noms, baseState, url, clientViewAuth, clientID)
+}
+
+// Pull pulls new server state from the client view via the diffserver.
+// TODO pass in auth (sandbox is hardcoded)
+func pull(noms types.ValueReadWriter, baseState Commit, url string, clientViewAuth string, clientID string) (Commit, servetypes.ClientViewInfo, error) {
+	baseMap := baseState.Data(noms)
+	pullReq, err := json.Marshal(servetypes.PullRequest{
+		ClientViewAuth: clientViewAuth,
+		ClientID:       clientID,
+		BaseStateID:    baseState.Meta.Snapshot.ServerStateID,
+		Checksum:       baseMap.Checksum(),
+	})
+	if err != nil {
+		return Commit{}, servetypes.ClientViewInfo{}, errors.New("could not marshal PullRequest")
+	}
+	verbose.Log("Pulling: %s from baseStateID %s", url, baseState.Meta.Snapshot.ServerStateID)
+	verbose.Log("Pulling: clientViewAuth: %s", clientViewAuth)
+
+	req, err := http.NewRequest("POST", url, bytes.NewReader(pullReq))
+	if err != nil {
+		return Commit{}, servetypes.ClientViewInfo{}, err
+	}
+	req.Header.Add("Authorization", sandboxAuthorization) // TODO expose this in the constructor so clients can set it
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return Commit{}, servetypes.ClientViewInfo{}, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		body, err := ioutil.ReadAll(resp.Body)
+		var s string
+		if err == nil {
+			s = string(body)
+		} else {
+			s = err.Error()
+		}
+		return Commit{}, servetypes.ClientViewInfo{}, fmt.Errorf("%s: %s", resp.Status, s)
+	}
+
+	var pullResp servetypes.PullResponse
+	var r io.Reader = resp.Body
+	err = json.NewDecoder(r).Decode(&pullResp)
+	if err != nil {
+		return Commit{}, servetypes.ClientViewInfo{}, fmt.Errorf("Response from %s is not valid JSON: %s", url, err.Error())
+	}
+
+	if pullResp.LastMutationID < baseState.Meta.Snapshot.LastMutationID {
+		return Commit{}, pullResp.ClientViewInfo, fmt.Errorf("Client view lastMutationID %d is < previous lastMutationID %d; ignoring", pullResp.LastMutationID, baseState.Meta.Snapshot.LastMutationID)
+	}
+	patchedMap, err := kv.ApplyPatch(noms, baseMap, pullResp.Patch)
+	if err != nil {
+		return Commit{}, pullResp.ClientViewInfo, errors.Wrap(err, "couldnt apply patch")
```

Comment:
i generally dont put quotes in error messages because it can make it visually harder to parse quoted values. but i agree this is a weird thing that i do so fixed :)